tests/test_ResolveTest.py

```diff
@@ -1,295 +1,283 @@
-                    ]
-                },
-            ),
-            # Test resolving multiple full citations to different documents
-            (
-                [full1, full3],
-                {
-                    Resource(full1): [full1],
-                    Resource(full3): [full3],
-                },
-            ),
-        ]
+    def checkResolution(
+        self, *expected_resolutions: Tuple[Optional[int], str]
+    ):
+        """Helper function to check how a list of citation strings is
+        resolved by resolve_citations().
 
-        for citations, resolution_dict in test_pairs:
-            with self.subTest(
-                "Testing citation resolution for %s..." % citations,
-                citations=citations,
-                resolution_dict=resolution_dict,
-            ):
-                self._assertResolution(citations, resolution_dict)
+        For example, suppose we want to check
+        resolutions for "1 U.S. 1. 1 U.S., at 2. 1 F.2d 1. 2 U.S., at 2.".
 
-    def test_supra_resolution(self):
-        test_pairs = [
-            # Test resolving a supra citation
-            (
-                [full4, supra1],
-                {
-                    Resource(full4): [full4, supra1],
-                },
-            ),
-            # Test resolving a supra citation when its antecedent guess matches
-            # two possible candidates. We expect the supra citation to not
-            # be resolved.
-            (
-                [full5, full6, supra2],
-                {
-                    Resource(full5): [full5],
-                    Resource(full6): [full6],
-                },
-            ),
-        ]
+        We can call this:
+            >>> self.checkResolution(
+            ...     (0, "1 U.S. 1."),
+            ...     (0, "1 U.S., at 2."),
+            ...     (1, "1 F.2d 1."),
+            ...     (None, "2 U.S., at 2."),
+            ... )
+        Meaning "1 U.S. 1." and "1 U.S., at 2." should resolve into the first
+        Resource, "1 F.2d 1." should resolve into the second Resource, and
+        "2 U.S., at 2." shouldn't be included in any Resource.
+
+        checkResolutionList converts the above input to expected_resolution_dict:
+            {
+                Resource(citation=<1 U.S. 1>): [<1 U.S. 1>, <1 U.S., at 2>],
+                Resource(citation=<1 F.2d 1>): [<1 F.2d 1>],
+            }
+
+        And then calls:
+            self.assertResolution(
+                [<1 U.S. 1>, <1 U.S., at 2>, <1 F.2d 1>, <2 U.S., at 2>],
+                expected_resolution_dict
+            )
+        """
+        # input we're building for self.assertResolution
+        expected_resolution_dict = defaultdict(list)
+        citations = []
+
+        # resources we've found so far
+        resources: List[Resource] = []
+
+        for i, cite_text in expected_resolutions:
+            # extract cite and make sure there's only one:
+            cites = get_citations(cite_text)
```

Comment:
This refactor gives me a bit of pause because to test `resolve_citations()` we now must assume that `get_citations()` is working properly. Of course, that's probably a fine assumption, but I've always thought that unit tests should try to atomically test the logic of one method at a time. Just something that stuck out to me, not a deal-breaker.

Reply:
What if instead of thinking of it as an overbroad unit test you think of it as an awesomely readable functional test? :)

I'll say that in practice I find the new thing easier to read/write/debug. E.g. comparing these two, I have an easier time confirming that the second one is testing what the comment says it is testing, and duplicating it to add new tests:

```
full6 = case_citation(reporter="F.2d", metadata={"plaintiff": "Ipsum"})
full9 = case_citation(reporter="F.2d", page="99", metadata={"defendant": "Ipsum"})
short5 = case_citation(reporter="F.2d", metadata={"antecedent_guess": "Ipsum"}, short=True)
...
            # Test resolving a short form citation when its reporter and
            # volume match two possible candidates, and when its antecedent
            # guess also matches multiple possibilities. We expect the short
            # form citation to not be resolved.
            (
                [full6, full9, short5],
                {
                    Resource(full6): [full6],
                    Resource(full9): [full9],
                },
            ),
```

becomes

```
        # Test resolving a short form citation when its reporter and
        # volume match two possible candidates, and when its antecedent
        # guess also matches multiple possibilities. We expect the short
        # form citation to not be resolved.
        self.checkResolution(
            (0, "Foo v. Bar, 1 U.S. 1."),
            (1, "Foo v. Bar, 1 U.S. 2."),
            (None, "Foo, 1 U.S., at 2."),
        )
```